source/common/upstream/health_checker_impl.cc

```diff
@@ -113,8 +109,14 @@ HttpHealthCheckerImpl::HttpHealthCheckerImpl(const Cluster& cluster,
       codec_client_type_(codecClientType(config.http_health_check().use_http2()
                                              ? envoy::type::HTTP2
                                              : config.http_health_check().codec_client_type())) {
+
   if (!config.http_health_check().service_name().empty()) {
-    service_name_ = config.http_health_check().service_name();
+    envoy::api::v2::core::HealthCheck mutable_config = config;
```

Comment:
nit: just create a matcher object here and populate it before emplacing it. Or emplace it directly into the optional and then set it.

Also, can you add a small comment about how the legacy field was a prefix match as that was never obvious.